Migrate Search Pipeline to SQLite FTS5 + Refactor Indexing Architecture

yanirmr opened the pull request:
This PR introduces a major upgrade to the search subsystem and indexing pipeline.
The previous implementation relied on LIKE '%query%' scans and custom UDFs, which caused full-table scans and resulted in slow query times.

This PR replaces that entire stack with an efficient SQLite FTS5–backed full-text index, improves metadata handling, restructures the index builder, and significantly reduces search latency.

Key Changes
1. Introduced FTS5 as the primary text index
- Added a new documents_fts virtual table using SQLite FTS5.
- Full transcript text is now stored only in FTS5 instead of the documents table.
- Implemented Unicode-aware tokenization (unicode61 remove_diacritics).
- Added an fts_doc_mapping table to map FTS rowids to our internal doc_id values.
- Added FTS optimization step after bulk insert.

2. Replaced LIKE/UDF-based search with FTS5-based search
- Exact search: uses FTS phrase queries + regex boundary verification.
- Partial search: uses prefix queries for narrowing + Python substring search.
- Regex search: extracts candidate tokens for FTS narrowing; falls back to full text only when unavoidable.

3. Migrated document storage to “metadata-only” table
- documents table now contains only metadata: uuid, source, episode, date, title.
- full_text moved entirely to FTS5.
- Updated TranscriptIndex.get_document_text() to read from FTS instead.

yanirmr committed:
Validating empty queries and changing default search mode to 'exact'.

yanirmr committed:
Refactor search function to validate empty queries after reading all filter parameters, preserving user preferences. Update pagination and error handling for improved user experience.

yanirmr committed:
Enhance CSV export functionality by implementing caching for search results.

yanirmr committed:
Add caching for search results to improve performance during exports.

yanirmr committed:
Handle force reindex - delete the database file

yanirmr committed:
Refactor DatabaseService and TranscriptIndex for FTS5 integration

- Removed user-defined functions for SQLite matching and replaced document storage with FTS5 for full-text search capabilities.
- Updated document retrieval methods in TranscriptIndex to utilize FTS5 for improved performance and accuracy.
- Enhanced database schema to include FTS5 virtual table and mapping for efficient document indexing.
- Implemented batch processing for document insertion, separating full text storage from metadata.
- Adjusted search methods to support exact, partial, and regex queries using FTS5 features.

yanirmr committed:
Update DatabaseService and IndexManager for improved SQLite performance

- Adjusted cache size in DatabaseService from 4GB to 512MB for optimized memory usage.
- Enhanced batch processing logic to dynamically calculate batch size based on the number of parameters per row, ensuring efficient data insertion.
- Updated IndexManager to reflect the global cache size setting in DatabaseService, removing redundant cache size configurations.
- Added cleanup logic in IndexManager to drop existing tables before schema setup, preventing duplicates and primary key errors.

yairl commented on app/routes/export.py:
```diff
@@ -49,12 +49,32 @@ def export_results_csv():
-    # Always perform a new search to get all results
-    logger.info(f"Performing new search for CSV export: {query} (mode: {search_mode}, filters: date_from={date_from}, date_to={date_to}, sources={sources})")
+    # Try to use cached search results first
+    cache_key = main.make_cache_key(query, search_mode, date_from, date_to, sources)
```
What's the purpose of caching?

Only use case I can think of is if you are passing between pages for the same query.

What did you have in mind?

yairl commented on app/services/db.py:
```diff
@@ -115,8 +67,10 @@ def batch_execute(self, sql: str, params_list: List[List[Any]]):
+        # Calculate batch size based on SQLite's 999 variable limit
+        # If each row has N parameters, we can insert floor(999/N) rows at most
+        params_per_row = len(params_list[0])
+        BATCH_SIZE = max(1, 999 // params_per_row)  # At least 1 row
```
Are you sure it's 999 for our version of SQLite?
Is it possible to get this value as a parameter from the Python version (as SQLite here is released as part of the Python build)?

yanirmr committed:
remove caching

yanirmr committed:
Introduced SQLITE_MAX_PARAMS constant to define the maximum number of bound parameters for SQLite, improving code clarity and maintainability.

yanirmr committed:
Refactor run.py and CLI for improved index management

- Added auto-build functionality to automatically create the index if the database does not exist.
- Updated argument parsing in run.py to include --auto-build flag.
- Simplified index manager initialization in app/__init__.py to load from an existing database.
- Enhanced CLI commands for building and checking the status of the search index, replacing click with argparse for better usability.

yanirmr committed:
Refactor SQLite parameter handling and database schema setup

- Introduced a dynamic method to query the maximum number of bound parameters for SQLite, replacing the static SQLITE_MAX_PARAMS constant.
- Updated the database schema creation to remove the 'IF NOT EXISTS' clause for the documents and documents_fts tables, ensuring a clean setup.

yairl committed:
Merge branch 'main' into fix-bug-empy-search
